File: packages/go/xyp/registry_test.go
package xyp

import (
	"encoding/json"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"slices"
	"testing"
)

func TestRegistryCoversEveryServiceInTheSpec(t *testing.T) {
	t.Parallel()
	path := filepath.Join("..", "..", "..", "spec", "services.json")
	payload, err := os.ReadFile(path) //nolint:gosec // a fixed path inside the repository
	if errors.Is(err, fs.ErrNotExist) {
		// A downloaded module zip holds only this module's own files.
		t.Skipf("%s is not part of this module", path)
	}
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var services []struct {
		OperationName string `json:"operationName"`
	}
	if err := json.Unmarshal(payload, &services); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}

	want := make([]string, 0, len(services))
	for _, service := range services {
		want = append(want, service.OperationName)
	}
	slices.Sort(want)
	got := make([]string, 0, len(operationEndpoints))
	for operation := range operationEndpoints {
		got = append(got, operation)
	}
	slices.Sort(got)

	if !slices.Equal(got, want) {
		t.Errorf("the registry has %d operations, the spec %d", len(got), len(want))
	}
}

func TestKnownNamespacesAreSeededFromTheCheckedInWSDLs(t *testing.T) {
	t.Parallel()
	namespace, found := knownNamespaces["citizen-1.5.0"]
	if !found || namespace != "http://citizen.xyp.gov.mn/" {
		t.Errorf("knownNamespaces[citizen-1.5.0] = %q, %v", namespace, found)
	}
	if _, found := knownNamespaces["insurance-1.5.0"]; found {
		t.Error("insurance-1.5.0 has no WSDL in the repository, so it must be learned at runtime")
	}
}

File: packages/java/src/test/java/io/github/baljinnyamday/xyp/RegistryTest.java
package io.github.baljinnyamday.xyp;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Set;
import java.util.TreeSet;
import org.junit.jupiter.api.Test;

class RegistryTest {

  @Test
  void coversEveryServiceInTheSpec() throws IOException {
    Path spec = Path.of("..", "..", "spec", "services.json");
    // A downloaded source jar holds only this package's own files.
    assumeTrue(Files.exists(spec), spec + " is not part of this checkout");
    Set<String> want = new TreeSet<>();
    for (JsonNode service : new ObjectMapper().readTree(spec.toFile())) {
      want.add(service.get("operationName").asText());
    }

    assertEquals(want, new TreeSet<>(Registry.OPERATION_ENDPOINTS.keySet()));
  }

  @Test
  void knownNamespacesAreSeededFromTheCheckedInWsdls() {
    assertEquals("http://citizen.xyp.gov.mn/", Registry.KNOWN_NAMESPACES.get("citizen-1.5.0"));
    assertEquals("http://meta.xyp.gov.mn/", Registry.KNOWN_NAMESPACES.get("meta-1.5.0"));
    assertFalse(
        Registry.KNOWN_NAMESPACES.containsKey("insurance-1.5.0"),
        "insurance-1.5.0 has no WSDL in the repository, so it must be learned at runtime");
    assertEquals(
        "citizen-1.5.0", Registry.OPERATION_ENDPOINTS.get("WS100101_getCitizenIDCardInfo"));
  }
}
